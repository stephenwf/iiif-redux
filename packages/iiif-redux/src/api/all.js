const getAllCollections = state => state.resources.collections;
const getAllExternalResources = state => state.resources.externalResources;
const getAllLayers = state => state.resources.layers;
const getAllServices = state => state.resources.services;
const getAllManifests = state => state.resources.manifests;
const getAllSequences = state => state.resources.sequences;
const getAllRanges = state => state.resources.ranges;
const getAllCanvases = state => state.resources.canvases;
const getAllCanvasReferences = state => state.resources.canvasReferences;
const getAllImages = state => state.resources.imageResources;
const getAllAnnotations = state => state.resources.annotations;
const getAllAnnotationLists = state => state.resources.annotationLists;
const getAllAnnotationPages = state => state.resources.annotationPages;
const getAllAnnotationCollections = state =>
  state.resources.annotationCollections;
const getAllContentResources = state => state.resources.contentResources;
const getAllResources = state => state.resources;

export {
  getAllCollections,
  getAllExternalResources,
  getAllLayers,
  getAllServices,
  getAllManifests,
  getAllSequences,
  getAllRanges,
  getAllCanvases,
  getAllCanvasReferences,
  getAllImages,
  getAllAnnotationLists,
  getAllAnnotationPages,
  getAllAnnotationCollections,
  getAllAnnotations,
  getAllContentResources,
  getAllResources,
};
