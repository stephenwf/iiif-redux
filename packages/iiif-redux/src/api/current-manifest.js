import { getCurrentManifest } from './current';
import manifest from './manifest';

const {
  // Technical
  getId,
  getType,
  getViewingHint,
  getViewingDirection,
  getNavDate,
  // Descriptive
  getLabel,
  getDescription,
  getMetadata,
  getAttribution,
  getLogo,
  getLicense,
  getThumbnailId,
  getThumbnail,
  // Linking
  getWithinIds,
  getWithin,
  getRenderingIds,
  getRendering,
  getRelatedIds,
  getRelated,
  getServiceIds,
  getService,
  getSeeAlsoIds,
  getSeeAlso,
  // Structural
  getSequenceIds,
  getSequences,
  getRangeIds,
  getRanges,
  getStructureIds,
  getStructures,
  getOtherContentIds,
  getOtherContent,
} = manifest(getCurrentManifest);

export {
  getCurrentManifest,
  // Technical
  getId,
  getType,
  getViewingHint,
  getViewingDirection,
  getNavDate,
  // Descriptive
  getLabel,
  getDescription,
  getMetadata,
  getAttribution,
  getLogo,
  getLicense,
  getThumbnailId,
  getThumbnail,
  // Linking
  getWithinIds,
  getWithin,
  getRenderingIds,
  getRendering,
  getRelatedIds,
  getRelated,
  getServiceIds,
  getService,
  getSeeAlsoIds,
  getSeeAlso,
  // Structural
  getSequenceIds,
  getSequences,
  getRangeIds,
  getRanges,
  getStructureIds,
  getStructures,
  getOtherContentIds,
  getOtherContent,
};
