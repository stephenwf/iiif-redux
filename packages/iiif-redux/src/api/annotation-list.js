import { getAllAnnotationLists } from './all';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2annotationList from './2.x/annotation-list';

const annotationList = presentation2annotationList;

export default annotationList;

export const annotationLists = resourceListSelectorFactory(
  getAllAnnotationLists,
  annotationList
);

export const annotationListByIdSelector = byIdSelectorFactory(
  annotationList,
  'annotationLists'
);
