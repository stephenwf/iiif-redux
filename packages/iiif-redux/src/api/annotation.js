import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2Annotation from './2.x/annotation';
import { getAllAnnotations } from './all';

const annotation = presentation2Annotation;

export default annotation;

export const annotations = resourceListSelectorFactory(
  getAllAnnotations,
  annotation
);

export const annotationByIdSelector = byIdSelectorFactory(
  annotation,
  'annotations'
);
