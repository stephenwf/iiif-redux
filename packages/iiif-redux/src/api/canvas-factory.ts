import { createSelector, OutputSelector } from 'reselect';
import * as technical from './iiif-technical';
import { TechnicalProperties } from './iiif-technical';
import { DescriptiveProperties, Language, Logo } from './iiif-descriptive';
import { getAllImages } from './all';
import { getCurrentCanvas } from './current';
import * as descriptive from './iiif-descriptive';

interface CanvasFactory {
  // Technical
  getId: OutputSelector<{}, string, (res: TechnicalProperties) => string>;
  getType: OutputSelector<{}, string, (res: TechnicalProperties) => string>;
  getViewingHint: OutputSelector<
    {},
    string | null,
    (res: TechnicalProperties) => string | null
  >;
  getHeight: OutputSelector<{}, number, (res: TechnicalProperties) => number>;
  getWidth: OutputSelector<{}, number, (res: TechnicalProperties) => number>;

  // Descriptive.
  getLabel: OutputSelector<
    {},
    Language,
    (res: DescriptiveProperties) => Language
  >;
  getDescription: OutputSelector<
    {},
    Language,
    (res: DescriptiveProperties) => Language
  >;
  getAttribution: OutputSelector<
    {},
    Language,
    (res: DescriptiveProperties) => Language
  >;
  getMetadata: OutputSelector<
    {},
    Language,
    (res: DescriptiveProperties) => Language
  >;
  getThumbnailId: OutputSelector<
    {},
    string | Array<string>,
    (res: DescriptiveProperties) => string | Array<string>
  >;
  getLicense: OutputSelector<
    {},
    Array<string>,
    (res: DescriptiveProperties) => Array<string>
  >;
  getLogo: OutputSelector<
    {},
    Logo | Array<Logo> | string | Array<string>,
    (res: DescriptiveProperties) => Logo | Array<Logo> | string | Array<string>
  >;
}

type CanvasSelector = () => any;

function createCanvasSelectors(canvasSelector: CanvasSelector): CanvasFactory {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   * - getHeight
   * - getWidth
   **************************************************/

  const getId = createSelector(canvasSelector, technical.getId);

  const getType = createSelector(canvasSelector, technical.getType);

  const getViewingHint = createSelector(
    canvasSelector,
    technical.getWhitelistedViewingHint(['non-paged', 'facing-pages'])
  );

  const getHeight = createSelector(canvasSelector, technical.getHeight);

  const getWidth = createSelector(canvasSelector, technical.getWidth);

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getMetadata
   * - getDescription
   * - getThumbnail
   * - getAttribution
   * - getLicense
   * - getLogo
   **************************************************/

  const getLabel = createSelector(getCurrentCanvas, descriptive.getLabel);

  const getDescription = createSelector(
    getCurrentCanvas,
    descriptive.getDescription
  );

  const getMetadata = createSelector(getCurrentCanvas, descriptive.getMetadata);

  const getAttribution = createSelector(
    getCurrentCanvas,
    descriptive.getAttribution
  );

  const getLogo = createSelector(getCurrentCanvas, descriptive.getLogo);

  const getLicense = createSelector(getCurrentCanvas, descriptive.getLicense);

  const getThumbnailId = createSelector(
    getCurrentCanvas,
    descriptive.getThumbnailId
  );

  return {
    getId,
    getType,
    getViewingHint,
    getHeight,
    getWidth,
    getLabel,
    getDescription,
    getMetadata,
    getAttribution,
    getLogo,
    getLicense,
    getThumbnailId,
  };
}

export default createCanvasSelectors;
