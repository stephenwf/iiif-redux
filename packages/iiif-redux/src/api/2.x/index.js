import annotationList, {
  mappings as annotationListMapping,
} from './annotation-list';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import externalResource, {
  mappings as externalResourceMapping,
} from './external-resource';
import imageResource, {
  mappings as imageResourceMapping,
} from './image-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';
import sequence, { mappings as sequenceMapping } from './sequence';

export const mappings = {
  annotationList: annotationListMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  externalResource: externalResourceMapping,
  imageResource: imageResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
  sequence: sequenceMapping,

  // Aliases.
  annotationPage: annotationListMapping,
  contentResources: imageResourceMapping, // @todo change when image/external get merged.
};

export {
  annotationCollection,
  annotationList,
  annotation,
  canvas,
  collection,
  externalResource,
  imageResource,
  manifest,
  range,
  sequence,
};
