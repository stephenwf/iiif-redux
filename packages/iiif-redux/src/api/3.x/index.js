import annotationCollection, {
  mappings as annotationCollectionMapping,
} from './annotation-collection';
import annotationPage, {
  mappings as annotationPageMapping,
} from './annotation-page';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import contentResource, {
  mappings as contentResourceMapping,
} from './content-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';

export const mappings = {
  annotationCollection: annotationCollectionMapping,
  annotationPage: annotationPageMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  contentResource: contentResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
};

export {
  annotationCollection,
  annotationPage,
  annotation,
  canvas,
  collection,
  contentResource,
  manifest,
  range,
};
