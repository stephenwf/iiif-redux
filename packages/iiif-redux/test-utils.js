import createStore from './src/createStore';
import {
  selectCanvas,
  selectManifest,
  selectSequence,
} from './src/spaces/routing';
import * as currentManifest from './src/api/current-manifest';
import * as currentSequence from './src/api/current-sequence';
import * as currentCanvas from './src/api/current-canvas';
import * as reducers from './src/reducers';
import { combineReducers } from 'redux';

export function waitForRequest(store, id) {
  return new Promise(resolve => {
    store.subscribe(() => {
      const newState = store.getState();
      if (
        newState.dereferenced &&
        newState.dereferenced[id] &&
        newState.dereferenced[id].loading === false
      ) {
        resolve();
      }
    });
  });
}

const reducer = combineReducers(reducers);
export function makeStateFromActions(...actions) {
  return [{ type: '@@redux/INIT' }, ...actions].reduce(reducer, {});
}

export async function createStoreAndImportManifest(manifestJson) {
  const id = manifestJson['@id'];
  const store = createStore();
  fetch.mockResponseOnce(JSON.stringify(manifestJson));

  const whenRequestFinishes = waitForRequest(store, id);

  store.dispatch(selectManifest({ id }));

  await whenRequestFinishes;

  const sequences = currentManifest.getSequenceIds(store.getState());
  store.dispatch(selectSequence({ id: sequences[0] }));

  const canvases = currentSequence.getCanvasIds(store.getState());
  store.dispatch(selectCanvas({ id: canvases[0] }));

  return store;
}
