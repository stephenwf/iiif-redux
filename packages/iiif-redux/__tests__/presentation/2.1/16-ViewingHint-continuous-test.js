import { createStructuredSelector } from 'reselect';
import testJson from '../../fixtures/presentation/2.1/16/manifest';
import * as currentManifest from '../../../src/api/current-manifest';
import { createStoreAndImportManifest } from '../../../test-utils';

describe('iiif/presentation/2.1/16 Manifest: ViewingHint: continuous', () => {
  global.fetch = require('jest-fetch-mock');

  it('should render correct viewing hint', async () => {
    const store = await createStoreAndImportManifest(testJson);
    const state = store.getState();

    expect(
      createStructuredSelector({
        viewingHint: currentManifest.getViewingHint,
      })(state)
    ).toMatchSnapshot();
  });
});
