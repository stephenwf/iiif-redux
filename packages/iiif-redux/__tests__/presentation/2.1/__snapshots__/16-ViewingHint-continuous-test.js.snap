// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iiif/presentation/2.1/16 Manifest: ViewingHint: continuous should render correct viewing hint 1`] = `
Object {
  "viewingHint": "continuous",
}
`;
